docs/book/content/types/objects/error_handling.md

```diff
@@ -169,3 +178,293 @@ The specified structured error information is included in the [`extensions`](htt
   ]
 }
 ```
+
+## Errors Backed by GraphQL's Schema
+
+Rust's model of errors can be adapted for GraphQL. Rust's panic is
+similar to a `FieldError`. The whole query is aborted and nothing can
+be extracted (except for error related information). Not all errors
+require this strict handling. Recoverable or partial errors can be put
+into the GraphQL scheme. This way the client knows what can usually
+happen. To implement this approach all errors must be partitioned into
+the two error classes. The difference to the model in Rust is that
+critical means, e.g., database error. On the other hand, recoverable
+errors can be fixed by the user itself. If the user changes the input,
+then the error might get away. The prime example for recoverable
+errors is input validation.
+
+The error class specifies where the error is seen. Critical errors are
+returned as `FieldErrors` (from the previous section). Non-critical
+errors are part of the GraphQL scheme. Similar to Rust, GraphQL allows
+to have a similar error model with unions (see Unions).
+
+### Example Input Validation (simple)
+
+In this example, a basic input validation is implemented with GraphQL
+types. Strings are used to identify the problematic field name. Errors
+for a particular field are also returned as a string. In this example,
+the string contains a localized error message. However, it is also
+possible to return a unique string identifier.
+
+```rust
+#[derive(juniper::GraphQLObject)]
+pub struct Item {
+    name: String,
+    quantity: i32,
+}
+
+#[derive(juniper::GraphQLObject)]
+pub struct ValidationError {
+    field: String,
+    message: String,
+}
+
+#[derive(juniper::GraphQLObject)]
+pub struct ValidationErrors {
+    errors: Vec<ValidationError>,
+}
+
+#[derive(juniper::GraphQLUnion)]
+pub enum GraphQLResult {
+    Ok(Item),
+    Err(ValidationErrors),
+}
+
+pub struct Mutation;
+
+#[juniper::graphql_object]
+impl Mutation {
+    fn addItem(&self, name: String, quantity: i32) -> GraphQLResult {
+        let mut errors = Vec::new();
+
+        if !(10 <= name.len() && name.len() <= 100) {
+            errors.push(ValidationError {
+                field: "name".to_string(),
+                message: "between 10 and 100".to_string()
+            });
+        }
+
+        if !(1 <= quantity && quantity <= 10) {
+            errors.push(ValidationError {
+                field: "quantity".to_string(),
+                message: "between 1 and 10".to_string()
+            });
+        }
+
+        if errors.is_empty() {
+            GraphQLResult::Ok(Item { name, quantity })
+        } else {
+            GraphQLResult::Err(errors)
+        }
+    }
+}
+
+```
+
+Each function may have a different return type. Depending on the input
+parameters a new result type is required. For example, adding a user
+require a new result type which contains the variant `Ok(User)`
+instead of `Ok(Item)`.
```

Comment:
```suggestion
Each function may have a different return type and depending on the input
parameters a new result type is required. For example, adding a user
requires a new result type which contains the variant `Ok(User)`
instead of `Ok(Item)`.
```